[#44856] Add submit action and upload functionality

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/44856
- added nextcloud upload client to frontend
- added upload workflow
- added error handling for upload

IMPORTANT: This PR does NOT fix the CORS issue we have connecting to nextcloud instances, neither the CSP issues. Hence, it is not functional on its own, but needs this two fixes to work.

To test it, there are workarounds to get both CORS and CSP working for local instances.

Commit:
[#44856] Get upload link from StorageFilesResourceService

Kharonus committed:
[#44856] Add submit action and upload functionality

- https://community.openproject.org/work_packages/44856
- added file selection input to storage component
- location picker is opened after file selection is complete

Kharonus committed:
[#44856] enabled file upload

- with hardcoded upload information
- no file link creation attached

Commit:
[#44856] Integrated frontend for upload with backend for share creation

Kharonus committed:
[#44856] added linking after upload and error handling

Kharonus committed:
[#44856] fixed file name encoding

Kharonus commented on lib_static/open_project/configuration/helpers.rb:
```diff
@@ -103,7 +103,8 @@ def remote_storage_hosts
           remote_storage_upload_host,
-          remote_storage_download_host
+          remote_storage_download_host,
+          "nextcloud.local"
```
TODO: Remove that

apfohl replied:
This change makes these spec fail:

```
rspec ./spec/requests/auth/oauth_login_csp_spec.rb:54 # CSP appends on login form from oauth with redirect-uri being a custom scheme appends given CSP appends from flash
rspec ./spec/requests/auth/oauth_login_csp_spec.rb:40 # CSP appends on login form from oauth when login required appends given CSP appends from flash
```

apfohl replied:
Removed.

Functionality now disabled. Will be active once https://community.openproject.org/projects/openproject/work_packages/45474/activity is done.

Kharonus committed:
Merge branch 'dev' into implementation/44856-add-submit-action-and-upload-functionality

Commit:
[#44856] StorageRepresenter returns multiple prepare upload links

Kharonus committed:
[#44856] use prepare link from storage response

Commit:
[#44856] Validate request body for prepare upload endpoint

Commit:
Merge remote-tracking branch 'origin/dev' into implementation/44856-add-submit-action-and-upload-functionality

Commit:
[#44856] Fixed specs

Commit:
[#44856] removed temporary information for CSP

Commit:
Merge remote-tracking branch 'origin/dev' into implementation/44856-add-submit-action-and-upload-functionality

Commit:
Merge remote-tracking branch 'origin/dev' into implementation/44856-add-submit-action-and-upload-functionality

Commit:
[#44856] Fixed ESLint problems

Commit:
Merge remote-tracking branch 'origin/dev' into implementation/44856-add-submit-action-and-upload-functionality

oliverguenther commented on frontend/src/app/core/state/storages/storage.model.ts:
```diff
@@ -27,14 +27,24 @@
+}
+
+export interface IPrepareUploadLink extends IHalMethodResourceLink {
+  payload:IPrepareUploadBody
```
```suggestion
  payload:IPrepareUploadBody;
```

Same as above

oliverguenther commented on frontend/src/app/core/state/storage-files/upload-link.model.ts:
```diff
@@ -0,0 +1,38 @@
+import { IHalMethodResourceLink, IHalResourceLinks } from 'core-app/core/state/hal-resource';
+
+interface IHalUploadResourceLinks extends IHalResourceLinks {
+  destination:IHalMethodResourceLink
```
```suggestion
  destination:IHalMethodResourceLink;
```

Even though eslint doesn't complain here (probably due to it being a single definition), please add semicolons to all interfaces. This makes editing later a bit easier with our current eslint setting

oliverguenther commented on frontend/src/app/shared/components/storages/file-picker-base-modal.component.ts/file-picker-base-modal.component.ts:
```diff
@@ -132,7 +132,7 @@ export abstract class FilePickerBaseModalComponent extends OpModalComponent impl
   }
 
-  private changeLevel(parent:string|null, crumbs:Breadcrumb[]):void {
+  protected changeLevel(parent:string | null, crumbs:Breadcrumb[]):void {
```
```suggestion
  protected changeLevel(parent:string|null, crumbs:Breadcrumb[]):void {
```

Our editor config sets no spaces between typedef union types, can you double check your editor that it applies these rules? Otherwise, we'll keep getting these changed back and forth

apfohl replied:
Interestingly, I changed it, because the build pipe discovered this..

oliverguenther commented on frontend/src/app/shared/components/storages/services/upload-storage-files.service.ts:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  private get propfindBody() {
+    return '<?xml version="1.0"?>\n'
+      + '<d:propfind xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns" xmlns:nc="http://nextcloud.org/ns">\n'
+      + '  <d:prop>\n'
+      + '    <oc:fileid />\n'
+      + '    <d:getlastmodified />\n'
+      + '    <d:getcontenttype />\n'
+      + '    <oc:size />\n'
+      + '    <oc:owner-display-name />\n'
+      + '  </d:prop>\n'
+      + '</d:propfind>';
```
A template literal would make this a bit easier to handle

```suggestion
    return `
    	<?xml version="1.0"?>
        <d:propfind xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns" xmlns:nc="http://nextcloud.org/ns">
          <d:prop>
            <oc:fileid />
            <d:getlastmodified />
            <d:getcontenttype />
            <oc:size />
            <oc:owner-display-name />
            </d:prop>
        </d:propfind>`;
```

oliverguenther reviewed (changes requested):
Some minor syntax remarks, otherwise mostly LGTM. I did not test the implementation due to lack of a local nextcloud

oliverguenther commented on frontend/src/app/shared/components/storages/services/upload-storage-files.service.ts:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  private parseXmlResponse(xml:string):IStorageFile {
+    const error = new Error(`Invalid response for uploaded file: ${xml}`);
```
Is there a specific reason you parse the XML through regexp instead of using something like a [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser)? I think that would have been cleaner and easier to read.

Kharonus replied:
replaced with DOMParser.

Kharonus committed:
Apply suggestions from code review

Co-authored-by: Oliver Günther <o.guenther@openproject.com>

Kharonus committed:
[#44856] replaced regex in upload response parsing

- use DOMParser instead of regex
- fixed linter for type unions and intersections

Kharonus committed:
Merge branch 'dev' into implementation/44856-add-submit-action-and-upload-functionality
